### ISSUE TITLE
[Code Addition Request]: Dynamic Hardware Resource Monitor with Prediction

### ISSUE THREAD

**Issue by SK8-infi** [UTSAVS26/PyVerse#1440]
### Have you completed your first issue?

- [x] I have completed my first issue

### Guidelines

- [x] I have read the guidelines
- [x] I have the link to my latest merged PR

### Latest Merged PR Link

https://github.com/UTSAVS26/PyVerse/pull/1438

### Project Description

---

# 🔮 Dynamic Hardware Resource Monitor with Prediction

A **Python-based, cross-platform system monitor** that not only displays real-time hardware usage but also uses machine learning to predict future CPU, GPU, and RAM usage for the next 60 seconds.

> Like `htop`, but smarter — with graphs, forecasts, and an interactive UI.

---

### 🔧 Real-Time Resource Monitoring
- CPU usage (per-core and total)
- RAM and Swap memory usage
- GPU load and VRAM usage (NVIDIA supported via `pynvml`)
- Disk I/O and usage
- Network I/O stats
- Temperature monitoring
- Battery status (if applicable)

### 📈 Visual Dashboard
- Live graphs for CPU, RAM, GPU usage
- Disk and network histograms
- Temperature visualizations (thermal gradient)
- Responsive dashboard using PyQt5 or Streamlit
- Export graphs as images

### 🤖 Forecasting with ML
- Predicts next 60 seconds of CPU, GPU, RAM usage
- Uses:
  - LSTM (via PyTorch/TensorFlow)
  - OR Statistical fallback (ARIMA, Exponential Smoothing)
- Real-time model training with sliding window of past 60 seconds
- Graph overlays of predicted vs actual usage

---

## 📂 Project Structure

```bash
hardware-monitor/
├── gui/
│   └── dashboard.py              # GUI logic (PyQt5/Streamlit)
├── sensors/
│   ├── cpu.py                    # CPU monitoring via psutil
│   ├── gpu.py                    # GPU stats (NVIDIA via pynvml)
│   └── memory.py                 # RAM and Swap
├── models/
│   ├── lstm_predictor.py         # LSTM-based forecasting
│   └── stats_fallback.py         # ARIMA/statistical fallback
├── visualizations/
│   ├── live_plot.py              # Real-time matplotlib/plotly graphs
│   └── prediction_overlay.py     # Forecast graph overlays
├── utils/
│   ├── logger.py                 # Save logs to file
│   └── config.py                 # Config/thresholds
├── main.py                       # Entry point
└── README.md
````

---

## 🧠 ML Prediction Details

* **Input**: Time series of CPU/GPU/RAM usage from past 60 seconds
* **Output**: 60-step forward forecast (1 value/sec)
* **Training**: On-the-fly or pre-trained; supports:

  * LSTM (Deep learning)
  * ARIMA / Holt-Winters (Stats-based fallback)
* **Model update frequency**: Every few minutes (configurable)

---

### 🔗 Dependencies

```bash
pip install psutil pynvml matplotlib plotly streamlit pyqt5 torch
```

### 💻 Run the Monitor

```bash
python main.py
```

Use `--mode=qt` or `--mode=web` for GUI preference.

---

## 📦 Optional Tools

* **Packaging**: Create portable executable with:

  ```bash
  pyinstaller --onefile main.py
  ```

* **Logging & Export**:

  * Export session stats as `.csv`
  * Export graphs as `.png`

---

## 🔮 Planned Features

* Plugin system for custom sensors
* Remote monitor via WebSockets
* Prometheus/InfluxDB integration
* Web-based dashboard for mobile view
* Auto-notifications on threshold breach

---

## 👨‍💻 Ideal For

* Developers running long training or compile jobs
* System admins benchmarking machines
* Students interested in time series forecasting
* Raspberry Pi or headless Linux box monitoring

---

## 📧 Contact

**Developer:** Shivansh Katiyar
**GitHub:** [[SK8-infi](https://github.com/SK8-infi)](https://github.com/SK8-infi)

---

### Full Name

Shivansh Katiyar

### Participant Role

SSOC
